File: src3/main/java/com/regenerationforrged/data/worldgen/preset/ContinentType.java
package com.regenerationforrged.data.worldgen.preset;

import com.mojang.serialization.Codec;

import net.minecraft.util.StringRepresentable;
import com.regenerationforrged.world.worldgen.GeneratorContext;
import com.regenerationforrged.world.worldgen.cell.continent.Continent;
import com.regenerationforrged.world.worldgen.cell.continent.advanced.AdvancedContinentGenerator;
import com.regenerationforrged.world.worldgen.cell.continent.fancy.FancyContinentGenerator;
import com.regenerationforrged.world.worldgen.cell.continent.infinite.InfiniteContinentGenerator;
import com.regenerationforrged.world.worldgen.cell.continent.simple.MultiContinentGenerator;
import com.regenerationforrged.world.worldgen.cell.continent.simple.SingleContinentGenerator;
import com.regenerationforrged.world.worldgen.util.Seed;

public enum ContinentType implements StringRepresentable {
    MULTI {
        
    	@Override
        public MultiContinentGenerator create(Seed seed, GeneratorContext context) {
            return new MultiContinentGenerator(seed, context);
        }
    }, 
    SINGLE {
        
    	@Override
        public SingleContinentGenerator create(Seed seed, GeneratorContext context) {
            return new SingleContinentGenerator(seed, context);
        }
    }, 
    MULTI_IMPROVED {
        
    	@Override
        public AdvancedContinentGenerator create(Seed seed, GeneratorContext context) {
            return new AdvancedContinentGenerator(seed, context);
        }
    }, 
    EXPERIMENTAL {
        
    	@Override
        public FancyContinentGenerator create(Seed seed, GeneratorContext context) {
            return new FancyContinentGenerator(seed, context);
        }
    },
    INFINITE {
        
    	@Override
        public InfiniteContinentGenerator create(Seed seed, GeneratorContext context) {
            return new InfiniteContinentGenerator(context);
        }
    };
	
	public static final Codec<ContinentType> CODEC = StringRepresentable.fromEnum(ContinentType::values);
    
    public abstract Continent create(Seed seed, GeneratorContext context);

    @Override
	public String getSerializedName() {
		return this.name();
	}
}

File: src2/main/kotlin/com/mbz/data/worldgen/preset/ContinentType.kt
package com.mbz.data.worldgen.preset

import com.mbz.world.worldgen.GeneratorContext
import com.mbz.world.worldgen.cell.continent.Continent
import com.mbz.world.worldgen.cell.continent.advanced.AdvancedContinentGenerator
import com.mbz.world.worldgen.cell.continent.fancy.FancyContinentGenerator
import com.mbz.world.worldgen.cell.continent.infinite.InfiniteContinentGenerator
import com.mbz.world.worldgen.cell.continent.simple.MultiContinentGenerator
import com.mbz.world.worldgen.cell.continent.simple.SingleContinentGenerator
import com.mbz.world.worldgen.util.Seed
import com.mojang.serialization.Codec
import net.minecraft.util.StringRepresentable

enum class ContinentType : StringRepresentable {
    MULTI {
        override fun create(seed: Seed, context: GeneratorContext) = MultiContinentGenerator(seed, context)
    },
    SINGLE {
        override fun create(seed: Seed, context: GeneratorContext) = SingleContinentGenerator(seed, context)
    },
    MULTI_IMPROVED {
        override fun create(seed: Seed, context: GeneratorContext) = AdvancedContinentGenerator(seed, context)
    },
    EXPERIMENTAL {
        override fun create(seed: Seed, context: GeneratorContext) = FancyContinentGenerator(seed, context)
    },
    INFINITE {
        override fun create(seed: Seed, context: GeneratorContext) = InfiniteContinentGenerator(context)
    };

    abstract fun create(seed: Seed, context: GeneratorContext): Continent

    override fun getSerializedName(): String = this.name

    companion object {
        @JvmField
        val CODEC: Codec<ContinentType> = StringRepresentable.fromEnum { values() }
    }
}
